draft-ietf-quic-tls.md

```diff
@@ -1496,10 +1496,11 @@ The endpoint that initiates a key update also updates the keys that it uses for
 receiving packets.  These keys will be needed to process packets the peer sends
 after updating.
 
-An endpoint SHOULD retain old keys so that packets sent by its peer prior to
-receiving the key update can be processed.  Discarding old keys too early can
-cause delayed packets to be discarded.  Discarding packets will be interpreted
-as packet loss by the peer and could adversely affect performance.
+An endpoint MUST retain old keys until it has successfully unprotected a packet
+sent using the new keys.  An endpoint SHOULD NOT discard old keys immediately
+after unprotecting a packet sent using the new keys.  Discarding old keys too
+early can cause delayed packets to be discarded.  Discarding packets will be
+interpreted as packet loss by the peer and could adversely affect performance.
```

Comment:
It *is* a performance issue provided the peer keeps sending.  You just lose some reordered packets and have spurious detection of some extra packet loss.  Where the connection hard-fails is when the peer never updates / learns about the update because the initiator stops sending.

Reply:
I'm not sure if I agree.

As @marten-seemann points out, an endpoint that only receives data (but does not send any) might decide to do a key update. If that endpoint drops the old receive key before any of the packets _it_ sent reaches the peer, the connection will be stuck.

That is because all the ack-eliciting packets involved will be the ones that are sent by the peer using the old key, which the endpoint has dropped that key.

Reply:
So here's what happens when you initiate a key update from key phase N to N+1: You start using N+1 for encrypting packets. As it's not guaranteed that those packets are received by the peer, it's not safe to drop keys for key phase N yet. You only know that the peer received a packet protected with N+1 keys once you successfully unprotect a packet encrypted with N+1 keys (as the peer is required to update its keys immediately once it receives your N+1 packet).
From this point on, it is **safe** to drop key N. However, it's not recommended, as there might be reordering, so keeping the key for a bit longer allows you to process delayed packets. At this point, this is just a performance optimization.

Reply:
Yes, it's safe to drop key N when you receive packets protected with key N+1.  And it's true that keeping keys for a short while helps avoid dropping reordered packets.

However, it it generally safer to only drop key N when you receive an acknowledgment for a packet sent with N+1 keys.  Because the other side might have updated on its own.  That means that the N+1 packet you received isn't any indication that they received your key update.  This distinction doesn't matter for dropping keys, but it does matter for determining whether you can initiate an update to N+2 keys later.

Reply:
> However, it it generally safer to only drop key N when you receive an acknowledgment for a packet sent with N+1 keys.

That's correct. In my mental model (and in my implementation), these are two separate conditions though. I can start the timer to drop the keys before I'm allowed to update to N+2.

Do you think we need to make any changes to this text here?

Reply:
I think that it's OK to have two events, and technically the drop events are bound to use of keys.  In that regard, I'm comfortable with the text you have.  I don't think we need to rely exclusively on acknowledgments for driving this.